tslearn/metrics/cysax.py

```diff
@@ -0,0 +1,150 @@
+STUFF_cysax = "cysax"
+
+import numpy as np
+from numba import njit, prange
+from sklearn.linear_model import LinearRegression
+
+__author__ = "Romain Tavenard romain.tavenard[at]univ-rennes2.fr"
+
+DTYPE = float
+DTYPE_INT = int
+
+
+@njit(parallel=True)
+def inv_transform_paa(dataset_paa, original_size):
+    n_ts = dataset_paa.shape[0]
+    sz = dataset_paa.shape[1]
+    d = dataset_paa.shape[2]
+    i = 0
+    t = 0
+    di = 0
+    t0 = 0
+    seg_sz = original_size // sz
+    dataset_out = np.zeros((n_ts, original_size, d))
+
+    for i in prange(n_ts):
+        for t in range(sz):
+            t0 = t * seg_sz
+            for di in range(d):
+                dataset_out[i, t0 : t0 + seg_sz, di] = dataset_paa[i, t, di]
+    return dataset_out
+
+
+@njit(parallel=True)
+def cydist_sax(sax1, sax2, breakpoints, original_size):
+    assert sax1.shape[0] == sax2.shape[0] and sax1.shape[1] == sax2.shape[1]
+    s = 0.0
+    sz = sax1.shape[0]
+    d = sax1.shape[1]
+    t = 0
+    di = 0
+    for t in prange(sz):
+        for di in range(d):
+            if np.abs(sax1[t, di] - sax2[t, di]) > 1:
+                max_symbol = max(sax1[t, di], sax2[t, di])
+                min_symbol = min(sax1[t, di], sax2[t, di])
```

Comment:
can we replace the inner for-loop with this:

```
mask = np.abs(sax1[t] - sax2[t]) > 1
max_sym = np.maximum(sax1[t, mask], sax2[t, mask])
min_sym = np.minimum(sax1[t, mask], sax2[t, mask])
s += np.square(breakpoints[max_symbol - 1] - - breakpoints[min_symbol])
```
Or, something like that?
if numba gives error due to advance indexing of `sax1[t, mask]`, then do: `sax1[t][mask]` (same for sax2)

Reply:
According to the previous cython file cysax.pyx, the input breakpoints is a numpy ndarray of dimension 1.
I do not think that we can write `breakpoints[max_symbol - 1]` or `breakpoints[min_symbol])`.

Reply:
this works for me:
```
breakpoints = np.arange(100)
max_symbol = np.array([2, 10, 4])
min_symbol = np.array([1, 7, 3])
print(breakpoints[max_symbol - 1])
print(breakpoints[min_symbol])

```
max_symbol is definitely integer. Otherwise, you cannot use it as index. So, `max_symbol - 1` is a mask that is used to get the corresponding element from 1D array `breakpoints`

Reply:
@YannCabanes have you tried this?

I suspect the last line should be 
```python
s += np.sum(np.square(breakpoints[max_symbol - 1] - breakpoints[min_symbol]))
```

but I might be wrong

Reply:
I have tried to replace the last loop by:

`mask = np.abs(sax1[t] - sax2[t]) > 1
max_sym = np.maximum(sax1[t, mask], sax2[t, mask])
min_sym = np.minimum(sax1[t, mask], sax2[t, mask])
s += np.sum(np.square(breakpoints[max_symbol - 1] - breakpoints[min_symbol]))`

It works with standard Python, but does not work with Numba.
I have the following error message related to `sax1[t, mask]` and `sax1[t, mask]`:
NumbaNotImplementedError: only one advanced index supported

I think that the same problem is discussed here:
https://stackoverflow.com/questions/70957881/numbanotimplementederror-only-one-advanced-index-supported-how-to-rewrite-a
The solution proposed is to use a loop instead of array indexing.